service: Support adding/removing a datacenter with tablets by changing RF

Deexie opened the pull request:
With this change, you can add or remove a DC(s) in a single ALTER KEYSPACE statement. It requires the keyspace to use rack list replication factor.

In existing approach, during RF change all tablet replicas are rebuilt at once. This isn't the case now. In global_topology_request::keyspace_rf_change the request is added to a ongoing_rf_changes - a new column in system.topology table. In a new column in system_schema.keyspaces - next_replication - we keep the target RF.

In make_rf_change_plan, load balancer schedules necessary migrations, considering the load of nodes and other pending tablet transitions. Requests from ongoing_rf_changes are processed concurrently, independently from one another. In each request racks are processed concurrently. No tablet replica will be removed until all required replicas are added. While adding replicas to each rack we always start with base tables and won't proceed with views until they are done (while removing - the other way around). The intermediary steps aren't reflected in schema. When the Rf change is finished:
- in system_schema.keyspaces:
  - next_replication is cleared;
  - new keyspace properties are saved;
- request is removed from ongoing_rf_changes;
- the request is marked as done in system.topology_requests.

Until the request is done, DESCRIBE KEYSPACE shows the replication_v2.

If a request hasn't started to remove replicas, it can be aborted using task manager. system.topology_requests::error is set (but the request isn't marked as done) and next_replication = replication_v2. This will be interpreted by load balancer, that will start the rollback of the request. After the rollback is done, we set the relevant system.topology_requests entry as done (failed), clear the request id from system.topology::ongoing_rf_changes, and remove next_replication.

Fixes: SCYLLADB-567.

No backport needed; new feature.

Deexie committed:
gms: add keyspace_multi_rf_change feature

mykaul commented:
This needs rebase and get back on track (if we aim to complete it for 2025.4)

Deexie commented:
This PR requires https://github.com/scylladb/scylladb/pull/25625 and needs to be adjusted for it. I will do it when that one is in soon-to-be-merged state. 

Note that the version here is strongly outdated and the code has been changed a lot since then.

Deexie commented:
> @Deexie how far are we from making this work (with rack-lists only) on top of #25625 and #24421?

Hard to tell, the version I currently have does not consider rack list at all. I will try to evaluate it. 

I pushed the current version so that I could navigate through it easier

Deexie committed:
db: service: add ongoing_rf_changes to system.topology

Following changes, will allow adding or removing all keyspace
replicas in a DC with a single ALTER KEYSPACE. For such operations,
the tablet load balancer needs to schedule rebuilds. To track
which RF change requests require rebuilds, we maintain a vector
of RF changes along with their ongoing rebuild phases.

Add a new ongoing_rf_changes column to system.topology to keep track
of those requests.

In this commit no data is kept in the new column.

Deexie committed:
db: add columns to system_schema.keyspaces

Add a new next_replication column to system_schema.keyspaces table.

While there is an ongoing RF change:
- next_replication keeps the target RF values;
- existing replication_v2 column keeps initial RF values - the ones we
  started the RF change with.

DESCRIBE KEYSPACE statement shows replication_v2.

When there is no ongoing RF change for this keyspace, its
next_replication is empty.

In this commit no data is kept in the new column.

Deexie committed:
service: add keyspace_rf_change_plan to migration_plan

Add keyspace_rf_change_plan to migration_plan.

The keyspace_rf_change_plan consists of:
- completion - info about the request for which all migrations are done. Only one
  request can be completed at the time, even if more have finished migrations
  (the rest will be completed later). Based on it:
    - next_replication is cleared;
    - new keyspace properties are saved (only if succeeded);
    - request is removed from ongoing_rf_changes;
    - the request is marked as done in system.topology_requests.
- aborts - info about requests that cannot complete because the required
  rf change is impossible (e.g. no available nodes in a required rack).
  Multiple requests can be aborted in a single plan. Based on each:
    - next_replication is set to current_replication (rolling back);
    - the request is marked as aborted with an error in system.topology_requests.

The scheduled rebuilds will be kept in migration_plan::_migrations.

Based on that the canonical_mutations are generated.

Add update_topology_state_with_mixed_change and use it if any schema
changes are required, i.e. if plan contains keyspace_rf_change_plan::completion.

Deexie committed:
service: split update_node_load_on_migration

Split update_node_load_on_migration into decrease_node_load and
increase_node_load - in the following changes for rebuilds we will
need only one of those at the time.

Deexie committed:
service: implement make_rf_change_plan

In make_rf_change_plan, load balancer schedules necessary migrations,
considering the load of nodes and other pending tablet transitions.
Requests from ongoing_rf_changes are processed concurrently, independently
from one another. In each request racks are processed concurrently.
No tablet replica will be removed until all required replicas are added.
While adding replicas to each rack we always start with base tables
and won't proceed with views until they are done (while removing - the other
way around).

Node availability is checked at two levels for extending actions:

1) In prepare_per_rack_rf_change_plan: the entire RF change request is
   aborted if any node in the target dc+rack is down, or if there are
   no live (non-excluded) nodes at all. Shrinking is never aborted.

2) In make_rf_change_plan: extending is skipped for a given round if
   any normal, non-excluded node in the target dc+rack is missing from
   the balanced node set. Shrinking always proceeds regardless.

The resulting behavior per node state combination (extending only):
  - all up                  -> proceed
  - some excluded + some up -> proceed (excluded nodes are skipped)
  - any down node           -> abort
  - all excluded (no live)  -> abort

When the last step is finished:
- in system_schema.keyspaces:
  - next_replication is cleared;
  - new keyspace properties are saved (if request succeeded);
- request is removed from ongoing_rf_changes;
- the request is marked as done in system.topology_requests.

Deexie committed:
service: rearrange keyspace_rf_change handler

In the following changes, keyspace_rf_change handler will also consider
a change of RF by more than one. Rearrange the handler, so that it
first chooses a kind of RF change and then creates relevant updates.

Do not wrap the code in schedule_migration function, as we no longer
need a quick return possibility.

Deexie committed:
service: handle multi_rf_change

Extend keyspace_rf_change handler to handle multi_rf_change.
multi_rf_change is allowed only if we add or remove DCs and
the keyspace uses rack list replication factor. The handler
adds the request id to topology::ongoing_rf_changes.
The request is further processed by load balancer.

Deexie committed:
cql3: allow changing RF by more than one when adding or removing a DC

rf_rack_valid_keyspaces relies on the fact that replicas of base
table and mv are streamed concurrently. This is no longer true
for newly introduced method of adding a DC. Disable rf_rack_valid_keyspaces
in test_mv_first_replica_in_dc to force the old method.

Deexie committed:
service: tasks: allow aborting ongoing RF changes

Allow aborting an ongoing RF change using task manager.

RF change can only be aborted if:
- it is currently paused (existing);
- it is a multi-RF change that still has replicas to be added.

In the second case, we set error for the request in system.topology_requests
and set next_replication to replication_v2. This makes load balancer
roll back the RF change.

Deexie committed:
test: test multi RF changes

Deexie committed:
docs: update documentation

bhalevy commented on docs/operating-scylla/procedures/cluster-management/decommissioning-data-center.rst:
```diff
@@ -102,3 +113,6 @@ Procedure
       UN  172.91.202.33  124.42 KB  256     32.6%             975edbm4-6564-63bd-san8-73fddce952ga   B1
+
+
+.. note:: Note that if `rf_rack_valid_keyspaces` option is set, there is no way to remove a DC with tablet keyspaces.
```
What should the user do in this case?
Let's describe the procedure how to migrate from `rf_rack_valid_keyspaces` to `enforce-rack-list`, and refer to it here and other places this caveat applies to.

Deexie committed:
service: fix indentation

Deexie commented:
@scylladbbot trigger-ci

annastuchlik reviewed (approved):
There are some minor suggestions, but overall, the documentation update looks good and was easy for me to follow. Thanks!

bhalevy commented:
@Deexie there are conflicts, please rebase

Deexie commented:
@tgrabiec please review

Deexie commented on docs/dev/system_schema_keyspace.md:
```diff
@@ -31,6 +34,11 @@ Columns:
+   If `previous_replication` is equal to `next_replication`, the keyspace rf change was aborted.
+* `next_replication` - the target replication factor for the keyspace during rf change.
+   If `next_replication` is equal to `replication_v2`, the keyspace rf change is finished.
+   If `next_replication` is equal to `previous_replication`, the keyspace rf change was aborted.
```
The sentences above are one-sided implications

Deexie commented:
> #### 🔴 CI State: FAILURE
> Mode 	Stage 	Status 	Queue Time 	Run Time 	Node
> 	[Checkout](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/26863//pipeline-overview) 	✅ 		2m 	spider9.cloudius-systems.com
> 	[Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/26863//pipeline-overview) 	✅ 		0m 	spider9.cloudius-systems.com
> 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/26863//pipeline-overview) 	✅ 	35m 	38m 	spider1.cloudius-systems.com
> dev 	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/26863//pipeline-overview) 	✅ 		14m 	spider1.cloudius-systems.com
> 	[Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/26863//pipeline-overview)
> 🔹 test_multi_rf_change
> 🔹 test_multi_rf_decrease_abort
> 🔹 test_multi_rf_increase_abort
> 🔹 test_multi_rf_increase_before_decrease
> 🔹 test_multi_rf_of_many_keyspaces
> 🔹 test/cql/cassandra_cql_test.result 	❌ 	0m 	2m 	spider1.cloudius-systems.com
> #### Build Details:
> 
>     * Total Queue: 35m
> 
>     * Total Run: 1h 32m
> 
> 
> **Note:** To re-trigger CI for this PR, comment: `@scylladbbot trigger-ci`

I have no idea how to handle that. It shows `[ERROR]: Unit Test failure`, but there is a message saying that all tests succeeded and no test logs. I found no further explanation of the failure. @yaronkaikov could you please help?

gusev-p commented on service/raft/group0_state_machine.cc:
```diff
@@ -298,8 +298,26 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
+        }
+        if (!other_mutations.empty()) {
+            co_await write_mutations_to_database(_ss, _sp, cmd.creator_addr, std::move(other_mutations));
+        }
```
> I assume I can't just happily move it there. Is there a way to migrate a table to schema commitlog?

I think you can just add `system.repair_tasks` to this list. When the user restarts a node after update, it will replay from the regular commitlog and start writing to the schema commitlog. If the node crashes and is rolled back to the previous version, we'll still replay the repair_tasks mutations from schema commitlog.

> Some of them also fail with database - Must call apply() on the owning shard (0 not in {1}), which I don't fully understand yet, but looking at logs, it is probably caused by the same table as it usually appears close to moving to end_repair

A theory: repair_task mutations are distributed via mixed_change, `execute_do_merge_schema` calls atomic version if db.local().apply which assumes the mutations belong to the local shard. The group0 tables should live on shard0 (because schema commitlog is only initialized on shard0). Actually, `enable_schema_commitlog()` turns the `use_null_sharder = true` on. If we just enable this option, the node restart will trigger resharding -- the table will be moved to shard0.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1433,8 +1466,312 @@ class load_balancer {
+        node_load_map nodes;
+        const locator::topology& topo = _tm->get_topology();
+        topo.for_each_node([&] (const locator::node& node) {
+            if (node.get_state() == locator::node::state::normal && !node.is_excluded()) {
```
`&& !n.is_draining()`

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1433,8 +1466,312 @@ class load_balancer {
+
+            std::unordered_set<global_tablet_id> rebuild_tablet_ids;
+            std::unordered_map<sstring, std::vector<sstring>> new_current_replication = ks.metadata()->strategy_options() | std::views::transform([] (const auto& pair) {
+                return std::make_pair(pair.first, std::get<rack_list>(pair.second));
```
Do we validate somewhere that this request is started only on keyspaces which use rack lists?

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1433,8 +1466,312 @@ class load_balancer {
+                };
+
+                // Ascending load heap of candidate target nodes.
+                std::make_heap(nodes_by_load_dst.begin(), nodes_by_load_dst.end(), nodes_dst_cmp);
```
I don't think you populated node load here. ensure_node() doesn't do it.

Also, if we have incomplete load stats, we should not make decisions.

I think it may be better to integrate this with per-rack plan-maker, which already does all of this, and so that we don't duplicate the logic.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1433,8 +1466,312 @@ class load_balancer {
+            }) | std::ranges::to<std::unordered_map<sstring, std::vector<sstring>>>();
+            bool needs_replication_update = false;
+            for (const auto& [dc, rack] : racks_to_work_on) {
+                locator::endpoint_dc_rack endpoint{dc, rack};
```
the name "endpoint" usually means "node". endpoint_dc_rack is "endpoint's dc and rack". It's usually called "location" in other places.

Using "endpoint" for this is confusing.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1433,8 +1466,312 @@ class load_balancer {
+                        if ((rf_change_state == rf_change_state::needs_extending && replica) ||
+                                ((rf_change_state == rf_change_state::needs_shrinking || rf_change_state == rf_change_state::rollback) && !replica)) {
+                            return make_ready_future<>();
+                        }
```
If we're shrinking, shouldn't we set all_done = false if the tablet is in transition and still has a pending replica in this rack? Otherwise, we will finalize the shrinking while there are pending replicas in this rack.

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1440,8 +1473,312 @@ class load_balancer {
+                    dc_to_racks[node_dc_rack.dc].push_back(node_dc_rack.rack);
+                }
+                res_prep.merge(rf_state == rf_change_state::needs_extending || rf_state == rf_change_state::rollback ?
+                    subtract_replication(dc_to_racks, replication) : subtract_replication(replication, dc_to_racks));
```
I think this logic of continuing the current step can lead to a deadlock. If we're rolling back because there's not enough capacity (rack was lost), we won't be able to complete the current step.

Deexie committed:
service: extend tablet_migration_info to handle rebuilds

Make tablet_migration_info::{src,dst} optional, so that it can be
reused by rebuild, for respectively leaving and pending replica.

Deexie commented:
@tgrabiec please re-review

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1437,6 +1458,324 @@ class load_balancer {
+                        };
+                        auto next = ti.replicas;
+                        next.push_back(pending_replica);
+                        auto mig_streaming_info = get_migration_streaming_info(topo, ti, tablet_transition_info{
```
There is an overload of get_migration_streaming_info() which accepts `tablet_migration_info`, so you don't have to fake `tablet_transition_info`, just pull `mig` construction here.

Deexie commented:
- rebase

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1443,6 +1464,354 @@ class load_balancer {
+                lblogger.warn("No target nodes available for RF change extending plan in dc {}, rack {}", dc, rack);
+            }
+            // Filter out extending actions since there are no target nodes.
+            // Shrinking actions can still proceed without target nodes.
```
Don't we want to finish extending before shrinking? 

I saw earlier that we abort RF change when there are no nodes. So how can we hit this path? Also, it's inconsistent with the earlier logic for `has_unreachable_rack`, which aborts in this case.

Deexie replied:
We can get here also if a node is down, but non-excluded (and there is no nodes to drain).
```
// Apply skiplist only when not draining.
// It's unsafe to move tablets to non-skip nodes as this can lead to node overload.
if (nodes_to_drain.empty()) {
  for (auto host_to_skip : _skiplist) {
     if (auto handle = nodes.extract(host_to_skip)) {
        auto& node = handle.mapped();
         lblogger.debug("Ignoring dead node {}: state={}", node.id, node.node->get_state());
     }
  }
}
```
We could consider the `_skiplist` above in `prepare_per_rack_rf_change_plan`. But I think we should rather avoid complicating the logic here and rely on the fact that a user has to do something with the down node anyway (or otherwise the coordinator would be blocked). If they exclude (remove) the node or restart it, we know how to proceed.

Deexie replied:
If we decide that we want to consider the skiplist here, it would also be a little problematic for the https://scylladb.atlassian.net/browse/SCYLLADB-1412 in a sense that:
- addition of a replica to dead rack
- rollback of a replica removal from a dead rack

would be handled in a different way (we would need to wait for user action for the latter).

tgrabiec replied:
How do you propose to consider the _skiplist?

For load balancing, we want to balance what we can, so it's fine to balance within the UP node set. But for replica allocation, we always should consider all members. Otherwise UP members would get overloaded.

If a member is DOWN, we should fail the RF change.

> We can get here also if a node is down, but non-excluded (and there is no nodes to drain).

Shouldn't we fail, or wait in this case? Otherwise, shrinking would execute before extending, opposite to what what the logic wants.

Deexie replied:
> How do you propose to consider the _skiplist?
> 

The problem is that `prepare_per_rack_rf_change_plan` considers all nodes (UP and DOWN) and `make_rf_change_plan` considers only UP nodes (because it uses `nodes` created for tablet migration).

Hence, we either have to:
- skip the `_skiplist` in `prepare_per_rack_rf_change_plan` (but only if there is no ongoing draining) and fail the request if there are only DOWN (or DOWN plus excluded) nodes in target rack;
- consider DOWN nodes in `make_rf_change_plan`. The latter is kind of done currently: while we don't see any `nodes`, we do not proceed with the request (addition).

I'm for the latter because of clarity reasons.

> For load balancing, we want to balance what we can, so it's fine to balance within the UP node set. But for replica allocation, we always should consider all members. Otherwise UP members would get overloaded.
> 
> If a member is DOWN, we should fail the RF change.

Can't we wait until the node is removed/brought back? We won't rollback the RF change until then anyway.

> 
> > We can get here also if a node is down, but non-excluded (and there is no nodes to drain).
> 
> Shouldn't we fail, or wait in this case? Otherwise, shrinking would execute before extending, opposite to what what the logic wants.

For one request we do either shrinking or extending, never mixed. So if at this point we are having both `actions` this means they are for different keyspaces

tgrabiec replied:
> consider DOWN nodes in make_rf_change_plan. The latter is kind of done currently: while we don't see any nodes, we do not proceed with the request (addition).

We need something stronger. We shouldn't proceed with allocation ignoring a DOWN node, even if some nodes in the rack are UP. We can block until they come up or get excluded, that's done automatically by tablet rebuild transition.

So I think we should skip making a plan for the rack if the working set of nodes doesn't contain all normal and not excluded members of the rack.

Note it's not just skiplist which drops nodes from the working set, it's also missing load stats:
```
                if (!_load_sketch->has_complete_data(host)) {
                    incomplete_host.emplace(host);
                    incomplete_count++;
                    nodes_i = nodes.erase(nodes_i);
                } else {
```

> Can't we wait until the node is removed/brought back? We won't rollback the RF change until then anyway. 

If we lose a node, the first step of recovery is to mark it as excluded. Bootstrap/replace will block behind active RF change.  That's a problem because when we lose a node, we don't want to continue with RF change over smaller capacity, that can overload the nodes. It's safer if we abort in this case. Rollback may need to wait, but that's fine.

There's also the case when someone ran RF change already in a bad state. It's better UX to fail immediately than block.  In that case, there's nothing to roll-back, so abort doesn't have to wait.

> For one request we do either shrinking or extending, never mixed. So if at this point we are having both actions this means they are for different keyspaces

I see. OK.

jsmolar commented on docs/operating-scylla/procedures/cluster-management/decommissioning-data-center.rst:
```diff
@@ -102,6 +102,33 @@ Procedure
 
          Consider :ref:`upgrading rf_rack_valid_keyspaces option to enforce_rack_list option <keyspace-rf-rack-valid-to-enforce-rack-list>` to ensure all tablet keyspaces use rack lists.
 
+   If the keyspace uses rack list replication, update the replication factor in one ALTER KEYSPACE statement.
```
This part is confusing for me. Plese clarify

e.g.
```
If the keyspace uses rack-aware replication, you can update the replication settings in a single ALTER KEYSPACE statement when adding and/or removing datacenters.

In that statement:

Existing datacenters must keep their current replication factor.
A new datacenter can be assigned a replication factor (i.e., 0 → N).
An existing datacenter can be removed (i.e., N → 0).

You cannot change the replication factor of any datacenter that is not being added or removed in the same statement.
```

jsmolar replied:
Also, placing this in a new subsection will improve readability.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -3203,20 +3203,77 @@ future<> storage_service::abort_paused_rf_change(utils::UUID request_id) {
+                        const auto& node_dc_rack = tm.get_topology().get_node(r.host).dc_rack();
+                        dc_to_racks[node_dc_rack.dc].push_back(node_dc_rack.rack);
+                    }
+                    auto diff = subtract_replication(next_replication, dc_to_racks);
```
> And I should check the error anyway

I don't see the "error" field being checked.
